inc/3rd-party/perfect-images/classes/PerfectImages.php

```diff
@@ -0,0 +1,364 @@
+	}
+
+	/**
+	 * Restore the optimized full-sized file and replace it by the original backup file.
+	 *
+	 * This is to have the original user-uploaded (rather than the optimized full-size) image is in place
+	 * when Perfect Images (re)generates any new images.
+	 *
+	 * @param int $media_id A media attachment ID.
+	 */
+	public function restore_originally_uploaded_image( int $media_id ) {
+		$media = new Media( $media_id );
+
+		$fullsize_path = $media->get_raw_fullsize_path();
+
+		if ( empty( $fullsize_path ) ) {
+			return;
+		}
+
+		$original_path = $media->get_original_path();
+
+		if ( empty( $original_path ) ) {
+			return;
+		}
+
+		/** If these are not the same, WP will rebuild the full-size from the original along with the thumbnails. */
+		if ( $fullsize_path !== $original_path ) {
+			return;
+		}
+
+		$backup_path = $media->get_raw_backup_path();
+
+		if ( empty( $backup_path ) ) {
+			return;
+		}
+
+		$data = new Data( $media );
+
+		if ( ! $data->is_optimized() ) {
+			return;
+		}
+
+		$tmp_file_path = $this->get_temporary_file_path( $fullsize_path );
+
+		if ( $this->filesystem->exists( $fullsize_path ) ) {
+			$this->filesystem->move( $fullsize_path, $tmp_file_path, true );
+		}
+
+		$this->filesystem->copy( $backup_path, $fullsize_path );
+	}
+
+	/**
+	 * Add a newly generated retina size that will need to be optimized.
+	 *
+	 * @hooked wr2x_retina_file_added
+	 *
+	 * @param int    $media_id    The media attachment ID.
+	 * @param string $retina_file The retina filename.
+	 * @param string $size_name   The size name.
+	 */
+	public function add_retina_size( int $media_id, string $retina_file, string $size_name ) {
+		$this->retina_sizes[] = [
+			'media_id'    => $media_id,
+			'retina_file' => $retina_file,
+			'size_name'   => $size_name,
+		];
+	}
+
+	/**
+	 * Optimize newly generated retina sizes.
+	 *
+	 * @hooked wr2x_generate_retina
+	 *
+	 * @param int $media_id The attachment id of the retina images to optimize.
+	 */
+	public function optimize_retina_sizes( int $media_id ) {
+		$sizes = [];
+
+		foreach ( $this->retina_sizes as $size ) {
+			if ( $media_id === $size['media_id'] ) {
+				$sizes[] = $size['size_name'] . '@2x';
+			}
+		}
+
+		if ( empty( $sizes ) ) {
+			return;
+		}
+
+		$process = new Process( new Data( new Media( $media_id ) ) );
+
+		$media_opt_level    = $process->get_data()->get_optimization_level();
+		$optimization_level = $media_opt_level ?: Imagify_Options::get_instance()->get( 'optimization_level' );
+
+		$process->optimize_sizes( $sizes, $optimization_level );
+	}
+
+	/**
+	 * Filter a Media's get_media_files() response to include retina size data.
+	 *
+	 * @hooked imagify_media_files
+	 *
+	 * @param array $sizes The Media's size data.
+	 *
+	 * @return array Sizes data that includes retina sizes.
+	 */
+	public function add_retina_sizes_meta( array $sizes ): array {
+		foreach ( $sizes as $size => $size_data ) {
+			$retina_path = wr2x_get_retina( $size_data['path'] );
```

Comment:
Maybe a function exists check before trying to use a 3rd party function? We never know